scripts/dev/gen_stub.php

```diff
@@ -0,0 +1,307 @@
+            }
+            continue;
+        }
+
+        throw new Exception("Unexpected node {$stmt->getType()}");
+    }
+
+    return $funcInfos;
+}
+
+function funcInfoToCode(FuncInfo $funcInfo): string {
+    $code = '';
+    if ($funcInfo->cond) {
+        $code .= "#if {$funcInfo->cond}\n";
+    }
+    if ($funcInfo->return->type) {
+        $returnType = $funcInfo->return->type;
+        if ($returnType->isBuiltin) {
+            $code .= sprintf(
+                "ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(arginfo_%s, %d, %d, %s, %d)\n",
+                $funcInfo->name, $funcInfo->return->byRef, $funcInfo->numRequiredArgs,
+                $returnType->toTypeCode(), $returnType->isNullable
+            );
+        } else {
+            throw new Exception("Not implemented");
+        }
+    } else {
+        $code .= sprintf(
+            "ZEND_BEGIN_ARG_INFO_EX(arginfo_%s, 0, %d, %d)\n",
+            $funcInfo->name, $funcInfo->return->byRef, $funcInfo->numRequiredArgs
+        );
+    }
+
+    foreach ($funcInfo->args as $argInfo) {
+        if ($argInfo->isVariadic) {
+            if ($argInfo->type) {
+                throw new Exception("Not implemented");
+            }
+            $code .= sprintf(
+                "\tZEND_ARG_VARIADIC_INFO(%d, %s)\n",
+                $argInfo->byRef, $argInfo->name
+            );
+        } else if ($argInfo->type) {
+            if ($argInfo->type->isBuiltin) {
+                $code .= sprintf(
+                    "\tZEND_ARG_TYPE_INFO(%d, %s, %s, %d)\n",
+                    $argInfo->byRef, $argInfo->name,
+                    $argInfo->type->toTypeCode(), $argInfo->type->isNullable
+                );
+            } else {
+                $code .= sprintf(
+                    "\tZEND_ARG_OBJ_INFO(%d, %s, %s, %d)\n",
+                    $argInfo->byRef, $argInfo->name,
+                    $argInfo->type->name, $argInfo->type->isNullable
+                );
+            }
+        } else {
+            $code .= sprintf("\tZEND_ARG_INFO(%d, %s)\n", $argInfo->byRef, $argInfo->name);
+        }
+    }
+
+    $code .= "ZEND_END_ARG_INFO()";
+    if ($funcInfo->cond) {
+        $code .= "\n#endif";
+    }
+    return $code;
+}
+
+/** @param FuncInfo[] $funcInfos */
+function generateArginfoCode(array $funcInfos): string {
+    $code = "/* This is a generated file, edit the .stub.php file instead. */\n\n";
+    return $code . implode("\n\n", array_map('funcInfoToCode', $funcInfos));
+}
+
+function initPhpParser() {
+    $version = "4.2.2";
+    $phpParserDir = __DIR__ . "/PHP-Parser-$version";
+    if (!is_dir($phpParserDir)) {
+        $cwd = getcwd();
+        chdir(__DIR__);
+        passthru("wget https://github.com/nikic/PHP-Parser/archive/v$version.tar.gz");
+        passthru("mkdir PHP-Parser-$version");
+        passthru("tar xvzf v$version.tar.gz -C PHP-Parser-$version --strip-components 1");
+        passthru("rm v$version.tar.gz");
```

Comment:
Use `unlink()` instead for (Windows) portability?

Reply:
Does the rest of this work on Windows? I can replace the wget with file_get_contents and mkdir with PHP mkdir, but the `tar` invocation doesn't really have a good alternative (or should that make use of ext/phar?)

Reply:
Yes, rest works fine with PHP SDK, which has a [couple of *nix utils](https://github.com/microsoft/php-sdk-binary-tools/tree/master/msys2/usr/bin).

Reply:
Great, that makes things easier. I've replaced the rm with unlink.